src/rgw/rgw_dedup.cc

```diff
@@ -0,0 +1,2363 @@
+
+      if (d_ctl.local_paused) {
+        ldpp_dout(dpp, 10) << __func__ << "::wait on d_ctl.local_paused" << dendl;
+        d_cond.wait(cond_lock, [this]{return !d_ctl.local_paused || d_ctl.should_stop() ;});
+      }
+
+      if (d_ctl.remote_paused) {
+        ldpp_dout(dpp, 10) << __func__ << "::wait on d_ctl.remote_paused" << dendl;
+        d_cond.wait(cond_lock, [this]{return !d_ctl.remote_paused || d_ctl.should_stop() || d_ctl.local_pause_req;});
+      }
+    } // while loop
+
+    ldpp_dout(dpp, 5) << "Dedup background thread resumed!" << dendl;
+  }
+
+  //---------------------------------------------------------------------------
+  static bool all_shards_completed(cluster *p_cluster,
+                                   librados::IoCtx &ioctx,
+                                   work_shard_t num_work_shards,
+                                   uint64_t *p_total_ingressed)
+  {
+    return p_cluster->all_work_shard_tokens_completed(ioctx, num_work_shards,
+                                                      p_total_ingressed);
+  }
+
+  //---------------------------------------------------------------------------
+  void Background::work_shards_barrier(work_shard_t num_work_shards)
+  {
+    // Wait for other worker to finish ingress step
+    uint64_t total_ingressed = 0;
+    while (!all_shards_completed(&d_cluster, d_dedup_cluster_ioctx, num_work_shards, &total_ingressed)) {
+      ldpp_dout(dpp, 10) << "Waiting for object ingress completion, ttl="
+                         << d_heart_beat_max_elapsed_sec << dendl;
+      std::unique_lock cond_lock(d_cond_mutex);
+      d_cond.wait(cond_lock, [this]{return d_ctl.should_stop() || d_ctl.should_pause();});
+      if (unlikely(d_ctl.should_pause())) {
+        handle_pause_req(__func__);
+      }
+      if (unlikely(d_ctl.should_stop())) {
+        return;
+      }
+    }
+
+    ldpp_dout(dpp, 10) << "\n\n==Object Ingress step was completed on all shards! ("
+                       << total_ingressed << ")==\n" << dendl;
+    if (unlikely(d_ctl.should_pause())) {
+      handle_pause_req(__func__);
+    }
+  }
+
+  //---------------------------------------------------------------------------
+  void Background::run()
+  {
+    // 256x8KB=2MB
+    const uint64_t PER_SHARD_BUFFER_SIZE = DISK_BLOCK_COUNT *sizeof(disk_block_t);
+    ldpp_dout(dpp, 20) <<__func__ << "::dedup::main loop" << dendl;
+
+    while (!d_ctl.shutdown_req) {
+      if (unlikely(d_ctl.should_pause())) {
+        handle_pause_req(__func__);
+        if (unlikely(d_ctl.should_stop())) {
+          ldpp_dout(dpp, 5) <<__func__ << "::stop req after a pause" << dendl;
+          d_ctl.dedup_exec = false;
+        }
+      }
+
+      if (d_ctl.dedup_exec) {
+        dedup_epoch_t epoch;
+        if (setup(&epoch) != 0) {
+          derr << "failed setup()" << dendl;
+          return;
+        }
+        work_shard_t num_work_shards = epoch.num_work_shards;
+        md5_shard_t  num_md5_shards  = epoch.num_md5_shards;
```

Comment:
you must verify that `num_md5_shards < MAX_MD5_SHARD`
the code may change when an epoch object exists, and this will probably cause a crash when you access `disk_block_array_t::d_disk_arr`